File: patterns/01-design-patterns/03-behavioral/05-chain-of-responsibility/go/chain_of_responsibility.go
package chainofresponsibility

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type Handler interface {
	Handle(map[string]any, *[]string) (string, error)
}
type auth struct{ next Handler }

func (h auth) Handle(r map[string]any, v *[]string) (string, error) {
	*v = append(*v, "auth")
	if r["token"] != "valid" {
		return "", &PatternError{code: "unauthenticated"}
	}
	return h.next.Handle(r, v)
}

type quota struct{ next Handler }

func (h quota) Handle(r map[string]any, v *[]string) (string, error) {
	*v = append(*v, "quota")
	if r["quota"].(float64) <= 0 {
		return "", &PatternError{code: "quota_exhausted"}
	}
	return h.next.Handle(r, v)
}

type execute struct{}

func (execute) Handle(r map[string]any, v *[]string) (string, error) {
	*v = append(*v, "execute")
	return "accepted:" + r["payload"].(string), nil
}
func Evaluate(input map[string]any) (any, error) {
	var chain Handler = auth{quota{execute{}}}
	raw, _ := input["requests"].([]any)
	responses := make([]map[string]any, 0, len(raw))
	for _, item := range raw {
		visited := []string{}
		decision, err := chain.Handle(item.(map[string]any), &visited)
		if err != nil {
			return nil, err
		}
		responses = append(responses, map[string]any{"decision": decision, "visited": visited})
	}
	return map[string]any{"responses": responses}, nil
}

File: patterns/01-design-patterns/03-behavioral/05-chain-of-responsibility/python/chain_of_responsibility.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class Handler:
    name=""
    def __init__(self,next_handler=None):self.next=next_handler
    def forward(self,request,visited):return self.next.handle(request,visited) if self.next else None
class Auth(Handler):
    name="auth"
    def handle(self,request,visited):
        visited.append(self.name)
        if request.get("token")!="valid":raise PatternError("unauthenticated")
        return self.forward(request,visited)
class Quota(Handler):
    name="quota"
    def handle(self,request,visited):
        visited.append(self.name)
        if request.get("quota",0)<=0:raise PatternError("quota_exhausted")
        return self.forward(request,visited)
class Execute(Handler):
    name="execute"
    def handle(self,request,visited):visited.append(self.name);return f"accepted:{request['payload']}"
def evaluate(input_data):
    chain=Auth(Quota(Execute()));responses=[]
    for request in input_data.get("requests",[]):
        visited=[];responses.append({"decision":chain.handle(request,visited),"visited":visited})
    return{"responses":responses}
